db/blip_sync_context.go

```diff
@@ -616,11 +616,42 @@ func (bsc *BlipSyncContext) sendNoRev(sender *blip.Sender, docID, revID string,
 
 // Pushes a revision body to the client
 func (bsc *BlipSyncContext) sendRevision(sender *blip.Sender, docID, revID string, seq SequenceID, knownRevs map[string]bool, maxHistory int, handleChangesResponseCollection *DatabaseCollectionWithUser, collectionIdx *int) error {
-	rev, err := handleChangesResponseCollection.GetRev(bsc.loggingCtx, docID, revID, true, nil)
-	if base.IsDocNotFoundError(err) {
-		return bsc.sendNoRev(sender, docID, revID, collectionIdx, seq, err)
-	} else if err != nil {
-		return fmt.Errorf("failed to GetRev for doc %s with rev %s: %w", base.UD(docID).Redact(), base.MD(revID).Redact(), err)
+	rev, originalErr := handleChangesResponseCollection.GetRev(bsc.loggingCtx, docID, revID, true, nil)
+
+	// set if we find an alternative revision to send in the event the originally requested rev is unavailable
+	var replacedRevID string
+
+	collectionCtx, collectionErr := bsc.collections.get(collectionIdx)
+	if collectionErr != nil {
+		return collectionErr
+	}
+
+	if base.IsDocNotFoundError(originalErr) {
+		if !collectionCtx.sendReplacementRevs {
+			return bsc.sendNoRev(sender, docID, revID, collectionIdx, seq, originalErr)
+		}
+
+		base.DebugfCtx(bsc.loggingCtx, base.KeySync, "Unavailable revision for %q %s - finding replacement: %v", base.UD(docID), revID, originalErr)
+
+		// try the active rev instead as a replacement
+		replacementRev, replacementRevErr := handleChangesResponseCollection.GetRev(bsc.loggingCtx, docID, "", true, nil)
+		if replacementRevErr != nil {
+			base.DebugfCtx(bsc.loggingCtx, base.KeySync, "Active revision unavailable for %q - sending original norev for %q: %v", base.UD(docID), revID, replacementRevErr)
+			return bsc.sendNoRev(sender, docID, revID, collectionIdx, seq, originalErr)
+		}
+
+		// if this is a filtered replication, ensure the replacement rev is in one of the filtered channels
+		// normal channel access checks are already applied in GetRev above
+		if collectionCtx.channels != nil && replacementRev.Channels.NumMatches(collectionCtx.channels) == 0 {
```

Comment:
Did you consider a 'HasMatch' function that will exit as soon as it finds a match?  Might be useful to stop counting as soon as we see a match, in cases where the document or replication has a very large number of channels.